### DISCUSSION TITLE
How can Deno.readDir be promisified?

### DISCUSSION THREAD

**Discussion by papiando** [denoland/deno#13258] [Q&A] [answered]
In all my code, I don't use async functions. Now, to perform a directory search, I need to use async and Promise doesn't work.
In my view, returning a Promise would be more effective, and you can await the result if necessary. Is there any design reason to do this? Is there any alternative?

**Comment by lucacasonato** [comment, marked as answer]
`Deno.readDir` returns an async iterator. This is very much like a promise, except that it can return more than one value.

This is done so that you can list large directories efficiently.

**Comment by papiando** [comment]
Thanks, Luca,
I understand that, but it could have returned a Promise to an iterator. I managed to solve it by using Deno.readDirSync and wrapping it in a promise. So I guess I answered my second question.
Happy New Year, all.

**Reply by crowlKats** [reply]
> I understand that, but it could have returned a Promise to an iterator

no, then it would have to collect all the entries and then resolve, that would be not ideal, especially in a scenario with a lot of entries in a folder. 

What you could do is:
```ts
async function readDir(path: string): Promise<Deno.DirEntry[]> {
  const entries = [];
  for await (const entry of Deno.readDir(path) {
    entries.push(entry);
  }
  return entries;
}
```
